rust/gateway/src/main.rs

```diff
@@ -1,30 +1,107 @@
     let cli = Cli::parse();
     setup_global_subscriber(layer::Identity::new());
 
     let device_id = get_device_id();
-    let mut session = Session::connect(
+    let (connect_url, private_key) = login_url(
+        Mode::Gateway,
         cli.common.url,
-        SecretString::from(cli.common.secret),
-        device_id,
-        CallbackHandler,
-    )
-    .unwrap();
-    tracing::info!("new_session");
+        SecretString::new(cli.common.secret),
+        device_id.clone(),
+    )?;
+
+    tokio::spawn(backoff::future::retry_notify(
+        ExponentialBackoffBuilder::default()
+            .with_max_elapsed_time(None)
+            .build(),
+        move || {
+            connect(
+                device_id.clone(),
+                private_key.clone(),
+                Secret::new(SecureUrl::from_url(connect_url.clone())),
+            )
+            .map_err(backoff::Error::transient)
+        },
+        |error, t: Duration| {
+            tracing::warn!(retry_in = ?t, "Error connecting to portal: {error}");
+        },
+    ));
 
-    block_on_ctrl_c();
+    tokio::signal::ctrl_c().await?;
 
-    session.disconnect(None);
     Ok(())
 }
 
+async fn connect(
+    device_id: String,
+    private_key: StaticSecret,
+    connect_url: Secret<SecureUrl>,
+) -> Result<Infallible> {
+    // Note: This is only needed because [`Tunnel`] does not (yet) have a synchronous, poll-like interface. If it would have, ICE candidates would be emitted as events and we could just hand them to the phoenix channel.
+    let (control_tx, control_rx) = tokio::sync::mpsc::channel(1);
+    let signaler = ControlSignaler::new(control_tx);
+    let tunnel = Arc::new(Tunnel::new(private_key, signaler, CallbackHandler).await?);
+
+    tracing::debug!("Attempting connection to portal...");
+
+    let mut channel = phoenix_channel::PhoenixChannel::connect(connect_url, device_id).await?;
+    channel.join("gateway", ());
+
+    let channel = loop {
+        match future::poll_fn(|cx| channel.poll(cx))
+            .await
+            .context("portal connection failed")?
+        {
+            phoenix_channel::Event::JoinedRoom { topic } if topic == "gateway" => {
+                tracing::info!("Joined gateway room on portal")
+            }
+            phoenix_channel::Event::InboundMessage {
+                topic,
+                msg: IngressMessages::Init(init),
+            } => {
+                tracing::info!("Received init message from portal on topic {topic}");
+
+                tunnel
+                    .set_interface(&init.interface)
+                    .await
+                    .context("Failed to set interface")?;
+
+                break channel;
+            }
+            other => {
+                tracing::debug!("Unhandled message from portal: {other:?}");
+            }
+        }
+    };
+
+    let mut eventloop = Eventloop::new(tunnel, control_rx, channel);
+
+    future::poll_fn(|cx| eventloop.poll(cx)).await
```

Comment:
> Just curious, the advantage of using `poll_fn` here instead of implementing a `Future` for `EventLoop` is that you can use `&mut self` for `poll` instead of `Pin<&mut self>`?

That and you can directly navigate with your IDE down the callstack!